core/station/impl/src/services/system.rs

```diff
@@ -575,69 +662,257 @@ mod init_canister_sync_handlers {
+        let mut add_request_policies = request_policies
+            .iter()
+            .map(|request_policy| {
+                let request_policy_id = request_policy
+                    .id
+                    .as_ref()
+                    .map(|id| HelperMapper::to_uuid(id.clone()).map(|uuid| *uuid.as_bytes()))
+                    .transpose();
+
+                let input = AddRequestPolicyOperationInput {
+                    specifier: request_policy.specifier.clone().into(),
+                    rule: request_policy.rule.clone().into(),
+                };
+
+                request_policy_id.map(|request_policy_id| (input, request_policy_id))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // sorting criteria:
+        // - if a policy depends on another policy, the dependent policy should be added first
+        // - keep the original order of the policies otherwise
+        add_request_policies.sort_by(|a, b| {
+            if let Some(a_id) = &a.1 {
+                if specifier_has_reference_to_policy_id(&b.0.specifier, a_id) {
+                    return Ordering::Less;
+                }
+            }
+            Ordering::Greater
+        });
+
+        for (input, request_policy_id) in add_request_policies {
+            REQUEST_POLICY_SERVICE.add_request_policy_with_id(input, request_policy_id)?;
+        }
+
+        Ok(())
+    }
+
+    // Registers the initial accounts of the canister during the canister initialization.
+    pub async fn set_initial_assets(assets: &[InitAssetInput]) -> Result<(), ApiError> {
+        let add_assets = assets
+            .iter()
+            .map(|asset| {
+                let input = AddAssetOperationInput {
+                    name: asset.name.clone(),
+                    blockchain: BlockchainMapper::to_blockchain(asset.blockchain.clone())
+                        .expect("Invalid blockchain"),
+                    standards: asset
+                        .standards
+                        .iter()
+                        .map(|standard| {
+                            BlockchainMapper::to_blockchain_standard(standard.clone())
+                                .expect("Invalid blockchain standard")
+                        })
+                        .collect(),
+                    decimals: asset.decimals,
+                    symbol: asset.symbol.clone(),
+                    metadata: asset.metadata.clone().into(),
+                };
+
+                (
+                    input,
+                    *HelperMapper::to_uuid(asset.id.clone())
+                        .expect("Invalid UUID")
+                        .as_bytes(),
+                )
+            })
+            .collect::<Vec<(AddAssetOperationInput, UUID)>>();
+
+        for (new_asset, with_asset_id) in add_assets {
+            match ASSET_SERVICE.create(new_asset, Some(with_asset_id)) {
+                Err(ApiError { code, details, .. }) if &code == "ALREADY_EXISTS" => {
```

Comment:
How can an asset exist upon canister initialization? Because it is specified multiple times in the input (this sounds like an invalid input)?

Reply:
Before this update the default assets (only ICP) were created even when installing with disaster recovery. This can now be removed since the default assets are only added if no initial entries are specified.